src/frontend/Typechecker.ml

```diff
@@ -34,23 +34,28 @@ let attach_warnings x = (x, List.rev !warnings)
 let model_name = ref ""
 let check_that_all_functions_have_definition = ref true
 
+type function_indicator =
+  | NotInFunction
+  | NonReturning
+  | Returning of UnsizedType.t
+
 (* Record structure holding flags and other markers about context to be
    used for error reporting. *)
 type context_flags_record =
   { current_block: Env.originblock
   ; in_toplevel_decl: bool
-  ; in_fun_def: bool
-  ; in_returning_fun_def: bool
+  ; in_fun: function_indicator
```

Comment:
Other `in_*` prefixed fields are of type `bool` (`in_fun_def` reads like "is in a function definition" and expects a yes-or-no answer). Can you name this `fun_return`?

I also wonder if the fields `in_rng_fun_def`, `in_lp_fun_def`, and `in_udf_dist_def` could be replaced with a single field of type `unit Fun_kind.suffix`.